data-structures/binary-heap/java/MaxHeap.java

```diff
@@ -0,0 +1,143 @@
+import java.util.NoSuchElementException;
+
+// An efficient data type to get the maximum element.
+public class MaxHeap <E extends Comparable<E>> {
+    // The heap is implemented as an array instead of a tree with nodes.
+    private final int DEFAULT_CAPACITY = 8;
+    private final int RESIZING_FACTOR = 2;
+    private E[] heap;
+    private int size;
+
+    // Creating a heap adding n elements, one at a time, has n log n time complexity.
+    public MaxHeap () {
+        heap = (E[]) new Comparable[DEFAULT_CAPACITY];
+        size = 0;
+        heap[0] = null;
+    }
+
+    // Creating a heap out of an array is more efficient, with only linear time complexity.
+    public MaxHeap (E[] array) {
+        size = array.length;
+        // Creates the heap with capacity for the entire array.
+        // If insert operations are expected, then 2 * size is better.
+        heap = (E[]) new Comparable[size + 1];
+        System.arraycopy(array, 0, heap, 1, size);
+        heap[0] = null;
+
+        // Starting from the parent of the last element, going through all parents, perform sink operations.
+        // This ensures no element is smaller than its children.
+        for (int i = size / 2; i >= 1; i--)
+            sink(i);
+    }
+
+    // Inserts an element and maintains the heap invariant.
+    public void insert(E e) {
+        // Resizes the array, if it lacks capacity.
+        if (size + 1 == heap.length) resize(size * RESIZING_FACTOR);

```

Comment:
I suggest you being consistent with your identation style. Some one-line 'if' statements are placed in the same line, and others are not.